statsmodels/sandbox/mice/mice.py

```diff
@@ -0,0 +1,1295 @@
+
+_mice_example_2 = """
+    >>> imp = mice.MICEData(data)
+    >>> fml = 'y ~ x1 + x2 + x3 + x4'
+    >>> mice = mice.MICE(fml, sm.OLS, imp)
+    >>> results = []
+    >>> for x in mice:
+    >>>     results.append(x)
+    >>>     if len(results) > 10:
+    >>>         break
+"""
+
+class MICE(object):
+
+    __doc__ = """\
+    Multiple Imputation with Chained Equations.
+
+    This class can be used to fit most Statsmodels models to data sets
+    with missing values.
+
+    Parameters
+    ----------
+    model_formula : string
+        The model formula to be fit to the imputed data sets.
+    model_class : statsmodels model
+        The model to be fit to the imputed data sets.
+    data : MICEData instance
+        MICEData object containing the data set for which
+        missing values will be imputed
+    n_skip : int
+        The number of imputed datasets to skip between consecutive
+        imputed datasets that are used for analysis.
+    init_kwds : dict-like
+        Dictionary of keyword arguments passed to the init method
+        of the analysis model.
+    fit_kwds : dict-like
+        Dictionary of keyword arguments passed to the fit method
+        of the analysis model.
+
+    Examples
+    --------
+    Simple example using defaults:
+    %(mice_example_1)s
+
+    Example using an iterator:
+    %(mice_example_2)s
+    """ % {'mice_example_1' : _mice_example_1,
+           'mice_example_2' : _mice_example_2}
+
+
+    def __init__(self, model_formula, model_class, data, n_skip=3,
+                 init_kwds=None, fit_kwds=None):
+
+        self.model_formula = model_formula
+        self.model_class = model_class
+        self.n_skip = n_skip
+        self.data = data
+        self.results_list = []
+
+        self.init_kwds = init_kwds if init_kwds is not None else {}
+        self.fit_kwds = fit_kwds if fit_kwds is not None else {}
+
+
+    def __iter__(self):
+        return self
+
+
+    def next(self):
+        return self.__next__()
+
+
+    def __next__(self):
+
+        self.data.update_all(self.n_skip)
+        start_params = None
+        if len(self.results_list) > 0:
+            start_params = self.results_list[-1].params
+        model = self.model_class.from_formula(self.model_formula,
+                                              self.data.data,
+                                              **self.init_kwds)
+        self.fit_kwds.update({"start_params": start_params})
+        result = model.fit(**self.fit_kwds)
+
+        return result
+
+
+    def fit(self, n_burnin=10, n_imputations=10):
+        """
+        Fit a model using MICE.
+
+        Parameters
+        ----------
+        n_burnin : int
+            The number of burn-in cycles to skip.
+        n_imputations : int
+            The number of data sets to impute
+        """
+
+        # Run without fitting the analysis model
+        self.data.update_all(n_burnin)
+
+        for j in range(n_imputations):
+            result = next(self)
```

Comment:
Do we need that the class itself is an iterator?

When Skipper and I took over stats.models we removed all iterators from the classes and converted them to iteration loops inside methods. The main problem is that it is very difficult to keep track what is actually changed and where the side effects are of the iterator.

I haven't looked at a specific example yet, this is just general skepticism to possibly difficult to understand code.